lxd/db/cluster/entities.go

```diff
@@ -16,13 +16,177 @@ import (
+	//    API/table). In other cases the caller may want to list all URLs of a particular type, so returning the ID along with
+	//    the URL allows for subsequent mapping or usage.
+	// 3. The project name (empty if the entity is not project specific).
+	// 4. The location (target) of the entity. Some entities require this parameter for uniqueness (e.g. storage volumes and buckets).
+	// 5. Path arguments that comprise the URL of the entity. These are returned as a JSON array in the order that they appear
+	//    in the URL.
+	AllURLs() string
+
+	// URLByID must return a SQL query that returns data in the same format as AllURLs. It must accept a bind argument
+	// that is the ID of the entity in the database.
+	URLByID() string
+
+	// URLsByProject must return a SQL query that returns data in the same format as AllURLs. It must accept a bind
+	// argument that is the name of the project that contains the entity.
+	URLsByProject() string
+
+	// IDFromURL must return a SQL query that returns the ID of an entity by the information contained in its unique URL in a common format.
+	// These queries are not used in isolation, they are used together as part of a larger UNION query.
+	// Because of this, all queries expect as arguments the project name, the location, and the path arguments of the URL.
+	// Some entity types don't require a project name or location, so that's why they explicitly check for an empty project
+	// name or location being passed in.
+	// Additionally, all of the queries accept an index number as their first binding argument so that the results can be correlated in
+	// the calling function (see PopulateEntityReferencesFromURLs below).
+	//
+	// TODO: We could avoid a query snippet per entity by making these snippets support multiple entities for a single entity type.
+	// (e.g. `WHERE projects.name IN (?, ...) AND instances.name IN (?, ...)` we'd need to be very careful!).
+	IDFromURL() string
+
+	// OnDeleteTriggerName must return the name of the trigger that runs when the entity type is deleted.
+	OnDeleteTriggerName() string
+
+	// OnDeleteTrigger must return SQL that creates a trigger with name OnDeleteTriggerName that runs when entities of
+	// the type are deleted from the database.
+	OnDeleteTrigger() string
+}
+
+// RequiresProject implements entity.Type for EntityType by calling RequiresProject on the underlying entityType. This
+// avoids a panic if the underlying entityType is nil (which is the case if the type code is entityTypeCodeNone).
+func (e EntityType) RequiresProject() bool {
+	if e.entityType == nil {
+		return false
+	}
+
+	return e.entityType.RequiresProject()
+}
+
+// Name implements entity.Type for EntityType by calling Name on the underlying entityType. This avoids a panic if the
+// underlying entityType is nil (which is the case if the type code is entityTypeCodeNone).
+func (e EntityType) Name() entity.TypeName {
+	if e.entityType == nil {
```

Comment:
Why do we need these nil checks, in what valid scenario would an EntityType be used with a nil e.entityType?
I'm worried it may mask an invalid state.

Reply:
When the `entityType` field of `EntityType` is nil, it behaves as the "none" entity type which is used by the warnings system. This made sense to be as being the zero value of `EntityType`.

Reply:
sounds like a good comment to have in the code base.

What does the warnings system do with a zero entity type btw?

Reply:
As far as I can tell there are four places where a warning is created that isn't specific to a particular entity (currently an empty string means the "none" entity type):
- https://github.com/canonical/lxd/blob/main/lxd/api_cluster.go#L2238 (unable to refresh cluster certificate)
- https://github.com/canonical/lxd/blob/main/lxd/daemon.go#L2058 (time skew on heartbeat)
- https://github.com/canonical/lxd/blob/main/lxd/daemon.go#L1639 (various DB warnings on startup)
- https://github.com/canonical/lxd/blob/main/lxd/daemon.go#L1604 (unable to connect to MAAS on init)

We could probably use the server entity type for these instead.

Reply:
I thought I had commented that the zero value was the "none" type. I'll update to make it clearer.

Reply:
I've updated the comments on `EntityType` to make it clear that the zero value of this type behaves as `entityTypeCodeNone`.

Reply:
> We could probably use the server entity type for these instead.

Yeah i think that makes sense.

Reply:
To clarify, would you like me to remove the concept of a "none" entity type altogether, and make the zero value of `EntityType` behave as the "server" type?

To do this, for compatibility when a `-1` (`entityTypeCodeNone`) is read from the database we'll treat it as a server entity type.

Reply:
I think that makes sense - do you?

Does "server" also represent "cluster wide" concept too?

Reply:
I think an entity of of "none" is confusing when it comes to granting entitlements.

Reply:
You can't grant entitlements against the "none" entity type as it's not in the authorization model. It's purely for warnings.

"server" represents cluster wide concepts so I think we should change it. This will eventually mean we can make the `entity_type_code` column in the warnings table non nullable but obviously we can't change that without a schema update.

Reply:
So I'll update `EntityType` so that the zero value is the "server" type :)

Reply:
yep thanks!

Reply:
I've updated `EntityType` such that it's zero-value is "server".